### PR TITLE
Improvements and additions

### PR DESCRIPTION
**Pull request by brandonsturgeon**
- Add Connection spam protection
 - Tweak spam detection values
 - Print more info about unrecognized messages
 - Cosmetic refactor of net handling
 - Break out of `net.Incoming` early if the message is likely/extreme spam
 - Track total message count along with specific message count

Example of the connection spam working
```
Client "Phatso" connected (172.17.10.1:27005).
L 03/25/2021 - 11:47:22: "Phatso<2><STEAM_0:0:21170873><>" STEAM USERID validated
L 03/25/2021 - 11:47:28: "Phatso<2><STEAM_0:0:21170873><>" disconnected (reason "Disconnect by user.")
Dropped Phatso from server (Disconnect by user.)
L 03/25/2021 - 11:47:29: "Phatso<3><STEAM_0:0:21170873><>" connected, address "172.17.10.1:27005"
Client "Phatso" connected (172.17.10.1:27005).
L 03/25/2021 - 11:47:29: "Phatso<3><STEAM_0:0:21170873><>" STEAM USERID validated
L 03/25/2021 - 11:47:30: "Phatso<3><STEAM_0:0:21170873><>" disconnected (reason "Disconnect by user.")
Dropped Phatso from server (Disconnect by user.)
L 03/25/2021 - 11:47:30: "Phatso<4><STEAM_0:0:21170873><>" connected, address "172.17.10.1:27005"
Client "Phatso" connected (172.17.10.1:27005).
L 03/25/2021 - 11:47:30: "Phatso<4><STEAM_0:0:21170873><>" STEAM USERID validated
L 03/25/2021 - 11:47:30: "Phatso<4><STEAM_0:0:21170873><>" disconnected (reason "Disconnect by user.")
Dropped Phatso from server (Disconnect by user.)
L 03/25/2021 - 11:47:31: "Phatso<5><STEAM_0:0:21170873><>" connected, address "172.17.10.1:27005"
Client "Phatso" connected (172.17.10.1:27005).
L 03/25/2021 - 11:47:31: "Phatso<5><STEAM_0:0:21170873><>" STEAM USERID validated
L 03/25/2021 - 11:47:31: "Phatso<5><STEAM_0:0:21170873><>" disconnected (reason "Disconnect by user.")
Dropped Phatso from server (Disconnect by user.)
L 03/25/2021 - 11:47:31: "Phatso<6><STEAM_0:0:21170873><>" connected, address "172.17.10.1:27005"
[CFC Section580] [warn] Spam connections from IP: 172.17.10.1:27005 - Banning: true
Client "Phatso" connected (172.17.10.1:27005).
L 03/25/2021 - 11:47:31: Addip: "<><><>" was banned by IP "for 5.00 minutes" by "Console" (IP "172.17.10.1:27005")
L 03/25/2021 - 11:47:31: "Phatso<6><STEAM_0:0:21170873><>" STEAM USERID validated
STEAMAUTH: Client Phatso received failure code 6
L 03/25/2021 - 11:47:32: "Phatso<6><STEAM_0:0:21170873><>" disconnected (reason "Client left game (Steam auth ticket has been canceled)
")
Dropped Phatso from server (Client left game (Steam auth ticket has been canceled)
)
```

### CODE REVIEW

File: moon/cfc_77.15.580/server/connect.moon [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    Section580.connectSpam[networkid] += 1
+
+    if Section580.connectSpam[networkid] > Section580.connectSpamThreshold
+        reason = "Too many connnections"
+
+        game.KickID userid, reason
+
+        return unless Section580.connectShouldBan
+
+        -- FIXME: Expect ULib at this point
```
Comment (samuelWilliams99) [inline moon/cfc_77.15.580/server/connect.moon:21]: is this not just 
`return unless ULib`
?

File: moon/cfc_77.15.580/server/connect.moon [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    connections = Section580.connectSpam[networkid]
+
+    if not connections
+        Section580.connectSpam[networkid] = 1
+        return
+
+    Section580.connectSpam[networkid] += 1
+
+    if Section580.connectSpam[networkid] > Section580.connectSpamThreshold
+        reason = "Too many connnections"
```
Comment (samuelWilliams99) [inline moon/cfc_77.15.580/server/connect.moon:15]: Could probably be a config value

File: moon/cfc_77.15.580/server/net.moon [hunk 3]
```diff
@@ -10,43 +10,71 @@ import flaggedMessages,
...
 net.Incoming = ( len, client ) ->
-    len -= 16
-    strName = NetworkIDToString ReadHeader!
+    header = ReadHeader!
+    strName = NetworkIDToString header
 
     return unless strName
+    if strName == "nil"
+        warnLog "Invalid network message sent by '#{client}': Header: '#{header}' | strName: '#{strName} | len: #{len}' - ignoring"
+        -- return TODO: Decide if we should RERO here
```
Comment (samuelWilliams99) [inline moon/cfc_77.15.580/server/net.moon:77]: We likely still want to tally it, but it'll later get ignored by the `if not func` and trigger a double warn

File: moon/cfc_77.15.580/server/net.moon [hunk 4]
```diff
@@ -59,14 +87,16 @@ net.Incoming = ( len, client ) ->
         plyNick = client\Nick!
         plyIP = client\IPAddress!
 
-    tallyUsage lowerStr, client, plySteamId, plyNick, plyIP
+    shouldIgnore = tallyUsage lowerStr, client, plySteamId, plyNick, plyIP
+    return unless shouldIgnore ~= nil
```
Comment (samuelWilliams99) [inline moon/cfc_77.15.580/server/net.moon:91]: why not just `return if shouldIgnore`?

**Review by samuelWilliams99 — APPROVED** [review]

File: moon/cfc_77.15.580/server/net.moon (hunk 4 above)
Comment (brandonsturgeon) [reply moon/cfc_77.15.580/server/net.moon:91]: Wasn't sure what I wanted `tallyUsage` to return

File: moon/cfc_77.15.580/server/connect.moon (hunk 1 above)
Comment (brandonsturgeon) [reply moon/cfc_77.15.580/server/connect.moon:21]: No, I definitely don't want this to silently do nothing. Either ULib exists and they get banned, or it doesn't and it errors

File: moon/cfc_77.15.580/server/net.moon (hunk 3 above)
Comment (brandonsturgeon) [reply moon/cfc_77.15.580/server/net.moon:77]: Weirdly enough, we don't see that functionality currently. When someone is actually spamming `nil` maliciously, only that alert triggers.

In either case, with `tallyUsage` instructing `net.Incoming` to stop processing a message, that shouldn't happen 👍

File: moon/cfc_77.15.580/server/connect.moon [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import \warnLog from Section580
```
Comment (plally) [inline moon/cfc_77.15.580/server/connect.moon:1]: you can do this?

Comment (brandonsturgeon) [reply moon/cfc_77.15.580/server/connect.moon:1]: yeah ❤️

**Pull request merged**